solver/app/v2/reject.go

```diff
@@ -14,12 +21,112 @@ const (
+// Errors are unexpected and refer to internal server problems.
+func newShouldRejector(
+	backends ethbackend.Backends,
+	solverAddr common.Address,
+	targetName func(Order) string,
+	chainName func(uint64) string,
+) func(ctx context.Context, srcChainID uint64, order Order) (rejectReason, bool, error) {
+	return func(ctx context.Context, srcChainID uint64, order Order) (rejectReason, bool, error) {
+		// reject swallows the error (only logging it) and returns true and the shouldReject reason.
+		reject := func(reason rejectReason, err error) (rejectReason, bool, error) {
+			err = errors.Wrap(err, "reject",
+				"order_id", order.ID.String(),
+				"dest_chain_id", order.DestinationChainID,
+				"src_chain_id", order.SourceChainID,
+				"target", targetName(order))
+
+			rejectedOrders.WithLabelValues(
+				chainName(order.SourceChainID),
+				chainName(order.DestinationChainID),
+				targetName(order),
+				reason.String(),
+			).Inc()
+
+			log.InfoErr(ctx, "Rejecting request", err, "reason", reason)
+
+			return reason, true, nil
+		}
+
+		// returnErr returns the error, with false and rejectNone. It should be used for unexpected errors.
+		returnErr := func(err error) (rejectReason, bool, error) {
+			return rejectNone, false, err
+		}
+
+		if srcChainID != order.SourceChainID {
+			return returnErr(errors.New("source chain id mismatch [BUG]", "got", order.SourceChainID, "expected", srcChainID))
+		}
+
+		destChainID := order.DestinationChainID
+		backend, err := backends.Backend(destChainID)
+		if err != nil {
+			return reject(rejectUnsupportedDestChain, err)
+		}
+
+		// check all expenses are supported, ethereum addressed tokens
+		var expenses []Expense
+		for _, output := range order.MaxSpent {
+			chainID := output.ChainId.Uint64()
+
+			// inbox contract order resolution should ensure output.chainId matches order.DestinationChainID (derived from fillInstructions)
+			if chainID != destChainID {
+				return returnErr(errors.New("max spent chain id mismatch [BUG]", "got", chainID, "expected", destChainID))
+			}
+
+			addr, err := cast.EthAddress(output.Token[:20])
+			if err != nil {
+				return returnErr(errors.Wrap(err, "cast token address"))
+			}
+
+			if !cmpAddrs(addr, output.Token) {
+				return reject(rejectUnsupportedToken, errors.New("non-eth addressed token", "addr", hexutil.Encode(output.Token[:])))
+			}
+
+			tkn, ok := tokens.find(chainID, addr)
+			if !ok {
+				return reject(rejectUnsupportedToken, errors.New("unsupported token", "addr", addr))
+			}
+
+			expenses = append(expenses, Expense{
+				token:  tkn,
+				amount: output.Amount,
+			})
+		}
+
+		// check liquidty, reject if insufficient
+		for _, expense := range expenses {
+			bal, err := expense.token.balanceOf(ctx, backend, solverAddr)
+			if err != nil {
+				return returnErr(errors.Wrap(err, "get balance", "token", expense.token.symbol))
+			}
+
+			sampleBalance(chainName(destChainID), expense.token, solverAddr, bal)
```

Comment:
avoid general monitoring inside logic, since it will not monitor if no orders, rather have periodic sampling of balances regardless of orders